core/services/gateway/connectionmanager.go

```diff
@@ -72,40 +117,155 @@ func (m *connectionManager) DONConnectionManager(donId string) DONConnectionMana
+				nodeState.conn = network.NewWSConnectionWrapper()
+				go donConnMgr.readLoop(nodeAddress, nodeState)
+			}
+			donConnMgr.closeWait.Add(len(donConnMgr.nodes))
+		}
 		return m.wsServer.Start(ctx)
 	})
 }
 
 func (m *connectionManager) Close() error {
 	return m.StopOnce("ConnectionManager", func() (err error) {
 		m.lggr.Info("closing connection manager")
-		return m.wsServer.Close()
+		err = multierr.Combine(err, m.wsServer.Close())
+		for _, donConnMgr := range m.dons {
+			close(donConnMgr.shutdownCh)
+			for _, nodeState := range donConnMgr.nodes {
+				nodeState.conn.Close()
+			}
+		}
+		for _, donConnMgr := range m.dons {
+			donConnMgr.closeWait.Wait()
+		}
+		return
 	})
 }
 
 func (m *connectionManager) StartHandshake(authHeader []byte) (attemptId string, challenge []byte, err error) {
 	m.lggr.Debug("StartHandshake")
-	// TODO (FUN-469): implement the handshake (extract node address, validate signature, create a new attempt, generate challenge)
-	return "", []byte{}, nil
+	nodeAddress, authHeaderElems, err := m.parseAuthHeader(authHeader)
+	if err != nil {
+		return "", nil, err
+	}
+	donConnMgr, ok := m.dons[authHeaderElems.DonId]
+	if !ok {
+		return "", nil, errors.New("invalid DON ID")
+	}
+	nodeState, ok := donConnMgr.nodes[nodeAddress]
+	if !ok {
+		return "", nil, errors.New("no such node")
+	}
+	nowTs := uint32(m.clock.Now().Unix())
+	ts := authHeaderElems.Timestamp
+	if ts < nowTs-m.config.AuthTimestampToleranceSec || nowTs+m.config.AuthTimestampToleranceSec < ts {
+		return "", nil, errors.New("timestamp out of tolerance zone")
+	}
+	nodeState.mu.RLock()
+	lastAcceptedTs := nodeState.lastAcceptedTs
+	nodeState.mu.RUnlock()
+	if ts <= lastAcceptedTs {
+		return "", nil, errors.New("timestamp too low")
+	}
+	attemptId, challenge, err = m.newAttempt(nodeState, nodeAddress, ts)
+	if err != nil {
+		return "", nil, err
+	}
+	return attemptId, challenge, nil
+}
+
+func (m *connectionManager) parseAuthHeader(authHeader []byte) (nodeAddress string, authHeaderElems *network.AuthHeaderElems, err error) {
+	n := len(authHeader)
+	if n < network.HandshakeAuthHeaderMinLen {
+		return "", nil, errors.New("auth header too short")
+	}
+	authHeaderElems, err = network.Unpack(authHeader[:n-network.HandshakeSignatureLen])
+	if err != nil {
+		return "", nil, errors.New("unable to parse auth header")
+	}
+	signature := authHeader[n-network.HandshakeSignatureLen:]
+	signer, err := ValidateSignature(signature, authHeader[:n-network.HandshakeSignatureLen])
+	nodeAddress = "0x" + hex.EncodeToString(signer)
+	return
+}
+
+func (m *connectionManager) newAttempt(nodeSt *nodeState, nodeAddress string, timestamp uint32) (string, []byte, error) {
+	challenge := make([]byte, m.config.AuthChallengeLen)
+	_, err := rand.Read(challenge)
+	if err != nil {
+		return "", nil, err
+	}
+	m.connAttemptsMu.Lock()
+	defer m.connAttemptsMu.Unlock()
+	m.connAttemptCounter++
+	newId := fmt.Sprintf("%s_%d", nodeAddress, m.connAttemptCounter)
```

Comment:
`map` can have struct{} keys as long as all struct's fields are primitive types - which is your case.
So you can do:
```
type key struct {
   nodeAddress string
   attemptCounter int64
}

mkey := key { nodeAddress: nodeAddress, attemptCounter: m.connAttemptCounter }
m.connAttempts[mkey] = ...
```
No need to `fmt.Sprintf`.

Reply:
What you do is fine, just less performant.